Add "getting started" doc and add some clarifying details to examples (#124)
justinlaughlin/stubs: docs/getting_started.md

```
Following these preliminary steps, we begin the basic workflow of SMART:
1. **Initialize SMART containers**: Each container (species, compartments, reactions, or parameters) is initialized by first creating individual objects of the class `smart.model_assembly.Species`, `smart.model_assembly.Compartment`, `smart.model_assembly.Reaction`, `smart.model_assembly.Parameter`. The initialization of these objects is described in some detail in Examples 1 and 3, or comprehensively in the API documentation.

    These SMART containers contain all instances of any given class. In general, the user should define an empty container of the correct type and then add all instances of each object manually. In this simple example, we consider a model with a single species (`A`) over a single surface (`surf`), with a single reaction (`r1`, with parameter `k` defining the rate of decay) corresponding to exponential decay over time:
    ```python
    surf = smart.model_assembly.Compartment("surf", 2, smart.units.unit.um, 1)
    A = smart.model_assembly.Species("A", 10, conc_unit, 1.0, D_unit, "surf")
    k = smart.model_assembly.Parameter("k", 1, 1/smart.units.unit.s)

>     r1 = smart.model_assembly.Reaction("r1", ["A"], [], param_map={"k":"k"}, eqn_f_str="k*A")

Reviewer: Maybe we should use a slightly different `param_map` than `{"k":"k"}`, as it is not clear what the key and item refer to (do they refer to the variable name or the name in the parameter initialization?)

The `Reaction` constructor takes in an equation written as a string, whose input arguments (`k` and `A`, is mapped to either a `Species`  or `Parameter`. The input argument in the `param_map` maps from the equation variable name to the name of the parameter, i.e.
```python
 species = smart.model_assembly.Species("spA", 10, conc_unit, 1.0, D_unit, "surf")
 parameter = smart.model_assembly.Parameter("prmk", 1, 1/smart.units.unit.s)
 reaction = smart.model_assembly.Reaction("r1", ["spA"], [], param_map={"k":"prmk"}, eqn_f_str="k*spA")
```
I don't necessarily think the names I suggest above is the most sensible, but something along these lines to distinguish between:
1. Variable name in python context
2. Name given to class-object through constructor
3. Name given in equations

Author: Agreed, do you think it is worth changing the parameter maps throughout the other examples for clarity, or is it sufficient to just have the clear example with explanation here? I've also thought we might consider allowing a user to bypass the parameter map if the parameter name in the equation matches the one given to the class object, maybe that could be a future feature.

Reviewer: I would only do it in this example, to illustrate the usage.

I don't think one should be able to bypass the parameter map (as it introduces alot of new possible errors)

    cc = smart.model_assembly.CompartmentContainer()
    cc.add([surf])
    sc = smart.model_assembly.SpeciesContainer()
    sc.add([A])
    pc = smart.model_assembly.ParameterContainer()
    pc.add([k])
    rc = smart.model_assembly.ReactionContainer()
    rc.add([r1])
    ```

> 2. **Generate/load in mesh**: Separately from defining the SMART containers, the user must either generate or load in a mesh. In Examples 2, 3, 4, and 6, this mesh is generated using gmsh, via our module `smart.mesh_tools`. Then meshio is used to convert the .msh file generated by gmsh into a mesh for use by DOLFIN (the computational backend of FEniCS). Alternatively, Examples 1 and 5 use built-in mesh functions in DOLFIN (either `dolfin.RectangleMesh()` or `dolfin.UnitCubeMesh()`). For some applications, the user may wish to generate a mesh using other software. These can then be loaded in for use by SMART by using meshio to convert to an XDMF file and loading into dolfin using `dolfin.XDMFFile()`. (see `smart.mesh_tools.gmsh_to_dolfin` for an example).

Reviewer: I would add a link to gmsh here, i.e.
```markdown
[Gmsh](https://gmsh.info/)
```
Similarly for 
```markdown
[meshio](https://github.com/nschloe/meshio)
[FEniCS](https://fenicsproject.org/)
```
I would also write "XDMF file" as 
```markdown
[XDMF-file](https://www.xdmf.org/index.php/XDMF_Model_and_Format)
```
 when it is not referring to the DOLFIN class

    Once the DOLFIN mesh and mesh functions have been defined, they should be written to file using the function `smart.mesh_tools.write_mesh`. This generates an hdf5 file that is readable by SMART. Finally, the `smart.mesh.ParentMesh` object is defined. In our simple example here, we generate a dolfin mesh using `dolfin.RectangleMesh`:
    ```python
    rect_mesh = dolfin.RectangleMesh(dolfin.Point(0.0, 0.0), dolfin.Point(1.0, 1.0), 10, 10)
    mf2 = dolfin.MeshFunction("size_t", rect_mesh, 2, 1)
    mf1 = dolfin.MeshFunction("size_t", rect_mesh, 1, 0)
    smart.mesh_tools.write_mesh(rect_mesh, mf1, mf2, "rect_mesh.h5")
    parent_mesh = smart.mesh.ParentMesh(mesh_filename="rect_mesh.h5", mesh_filetype="hdf5", name="parent_mesh")
    ```

3. **Initialize model object**: Once the SMART containers and the parent mesh have been defined, these can all be used to assemble a single `smart.model` object. In our example, we initialize the model object with the default configuration:
    ```python
    config_cur = smart.config.Config()
    model_cur = smart.model.Model(pc, sc, cc, rc, config_cur, parent_mesh)
    ```

> 4. **Specify solver configuration**: SMART solves the assembled reaction-transport equations using the finite element method. This finite element system is assembled using the FEniCS software package and is then solved using SNES in PETSc4Py. The user can provide problem-tailored solver specifications by calling `config_cur.solver.update()` (given a config object, `config_cur`). Minimally, the user should provide a time step and final time for the simulation:

Reviewer: I would refer to the SNES user guide here:
i.e. SNES ->
```markdown
[SNES](https://petsc.org/release/manual/snes/#ch-snes)
```

    ```python
    config_cur.solver.update({"final_t": 10.0, "initial_dt": 0.1})
    ```
    In this manner, the user can also specify "time_precision" (number of digits past the decimal point to round time, default is 6) and "print_assembly" (prints information during assembly process, default is True).
5. **Initialize variational problem and solver**: Once the model object has been initialized and the config object has been set, we can initialize the variational problem in FEniCS and initialize the SNES solver in PETSc. This is all done internally by calling the single function:
    ```python
    model_cur.initialize()
    ```

>     Advanced users may wish to alter the default SNES preconditioners and solvers used in SMART. An example of how to do so is included in Example 6; generally, this involves calling functions within the `smart.model.Model.solver` object. For instance, to change the solver tolerance, we can write `model_cur.solver.setTolerances(rtol=1e-4)`.

Reviewer: You can add:

See [the SNES API](https://www.mcs.anl.gov/petsc/petsc4py-current/docs/apiref/petsc4py.PETSc.SNES-class.html)
for more information about the SNES class.

> 6. **Start simulation**: Finally, we can start the simulation, writing the data at each time step. We save the initial data to an XDMF file before entering a loop to solve the system until the current time is greater than the specified final time. We also initialize a list to store the average concentration of A over time, then integrate over the surface (using DOLFIN) at each time step to compute the average values.

Reviewer: I would remove the word `finally`, as it seems like it then is an ordeal to get to the point of starting the simulation. Just say:
6. **Start simulation**:  Initially, data is saved to an XDMF-file entering a loop to solve the system until the current time is greater than the specified final time. At each time step, the resulting distribution of  `A` is stored to file, and the average of each species (in this case just `A`) is stored to a list that can be accessed after the loop.

One thing that stands out looking at this code snippet is the usage of `Species.u["u"]` which we do not really explain the reasoning for. Why is it called `u`? 
Why is it a dictionary? Does the dictionary only contain "u"?
- How do you know the name of the solution?

Author: Very good point. `Species.u` is a dictionary with two entries: `"u"` (current solution) and `"n"` (solution at previous time step). Makes sense for use within the modules, but would be better if the user didn't have to interact with this directly. Maybe we can also store the solution in a more obvious place, like `Species.sol`, for ease of accessibility in cases like this. Thoughts?

Reviewer: Should there be any other entries in this dictionary?
If it is always just two entries, we could instead store them as properties of the module.

Author: There are only two entries in the dictionary. To be sure I am interpreting correctly, do you mean perhaps we just have `Species.u` and `Species.n` instead of `Species.u["u"]` and `Species.u["n"]`? I think we could change that, it would just involve some other minor tweaks throughout the code.

Reviewer: We can wait with changing it from a dictionary to a proper property. Long-term, we should do this for the maintainability of the code (and also clarity of what the class consists of.

    ```python
    results = dolfin.XDMFFile(model_cur.mpi_comm_world, "A.xdmf")
    results.parameters["flush_output"] = True
    results.write(model_cur.sc["A"].u["u"], model_cur.t)
    avg_A = [A.initial_condition]
    dx = dolfin.Measure("dx", domain=model_cur.cc["surf"].dolfin_mesh)
    volume = dolfin.assemble(1.0*dx)
    while True:
        model_cur.monolithic_solve()
        results.write(model_cur.sc["A"].u["u"], model_cur.t)
        int_val = dolfin.assemble(model_cur.sc["A"].u["u"]*dx)
        avg_A.append(int_val / volume)
        if model_cur.t >= model_cur.final_t:
            break
    ```

> 7. **End simulation and visualize data**: We recommend writing data to XDMF files as shown in the previous step. These can then be readily visualized and postprocessed in Paraview. An example screenshot of this minimal example in Paraview is shown below.

Reviewer: ```suggestion
7. **End simulation and visualize data**: We recommend writing data to XDMF-files as shown in the previous step. These can then be readily visualized and post-processed in [Paraview](https://www.paraview.org/). An example screenshot of this minimal example in Paraview is shown below.
```

# Getting started with SMART

The basic SMART workflow is summarized in the diagram below. This is the general sequence of events followed in all of the SMART demos, which introduce applications of the software at increasing levels of complexity. Examples 1 and 2 give 2-dimensional cases, whereas examples 3-6 give increasingly complicated 3-dimensional cell signaling models.

![SMART flow chart](images/SMART_flow_chart.png)

> To use SMART, the user should plan to write a python script (or Jupyter notebook) that goes through the outlined steps. The easiest way to get started with this is to follow the template of the SMART demo Jupyter notebooks or .py files. In the `smart` and `smart-lab` images, these files exist in the `/repo/examples` folder.

Reviewer: ```suggestion
To use SMART, the user should plan to write a Python script (or Jupyter notebook) that goes through the outlined steps. The easiest way to get started with this is to follow the template of the SMART demo Jupyter notebooks (`.ipynb`) or `.py` files. In the `smart` and `smart-lab` images, these files exist in the `/repo/examples` folder.
```

    ```python
    surf_var = smart.model_assembly.Compartment("surf", 2, smart.units.unit.um, 1)
    A_var = smart.model_assembly.Species("A", 10, conc_unit, 1.0, D_unit, "surf")
    kdecay_var = smart.model_assembly.Parameter("kdecay", 1, 1/smart.units.unit.s)
    r1_var = smart.model_assembly.Reaction("r1", ["A"], [], param_map={"k":"kdecay"},
                                            species_map={"sp":"A"}, eqn_f_str="k*sp")
    cc = smart.model_assembly.CompartmentContainer()
    cc.add([surf_var])
    sc = smart.model_assembly.SpeciesContainer()
    sc.add([A_var])
    pc = smart.model_assembly.ParameterContainer()
    pc.add([kdecay_var])
    rc = smart.model_assembly.ReactionContainer()
    rc.add([r1_var])
    ```

> 2. **Generate/load in mesh**: Separately from defining the SMART containers, the user must either generate or load in a mesh. In Examples 2, 3, 4, and 6, this mesh is generated using [Gmsh](https://gmsh.info/), via our module `smart.mesh_tools`. Then [meshio](https://github.com/nschloe/meshio) is used to convert the .msh file generated by [Gmsh](https://gmsh.info/) into a mesh for use by DOLFIN (the computational backend of [FEniCS](https://fenicsproject.org/)). Alternatively, Examples 1 and 5 use built-in mesh functions in DOLFIN (either `dolfin.RectangleMesh()` or `dolfin.UnitCubeMesh()`). For some applications, the user may wish to generate a mesh using other software. These can then be loaded in for use by SMART by using [meshio](https://github.com/nschloe/meshio) to convert to an [XDMF-file](https://www.xdmf.org/index.php/XDMF_Model_and_Format) and loading into dolfin using `dolfin.XDMFFile()`. (see `smart.mesh_tools.gmsh_to_dolfin` for an example).

Reviewer: ```suggestion
2. **Generate/load in mesh**: Separately from defining the SMART containers, the user must either generate or load in a mesh. In Examples 2, 3, 4, and 6, this mesh is generated using [Gmsh](https://gmsh.info/), via our module `smart.mesh_tools`. Then [meshio](https://github.com/nschloe/meshio) is used to convert the `.msh` file generated by [Gmsh](https://gmsh.info/) into a mesh for use by DOLFIN (the computational backend of [FEniCS](https://fenicsproject.org/)). Alternatively, Examples 1 and 5 use built-in mesh functions in DOLFIN (either `dolfin.RectangleMesh()` or `dolfin.UnitCubeMesh()`). For some applications, the user may wish to generate a mesh using other software. These can then be loaded in for use by SMART by using [meshio](https://github.com/nschloe/meshio) to convert to an [XDMF-file](https://www.xdmf.org/index.php/XDMF_Model_and_Format) and loading into dolfin using `dolfin.XDMFFile()`. (see `smart.mesh_tools.gmsh_to_dolfin` for an example).
```

    smart.mesh_tools.write_mesh(rect_mesh, mf1, mf2, "rect_mesh.h5")
    parent_mesh = smart.mesh.ParentMesh(mesh_filename="rect_mesh.h5", mesh_filetype="hdf5", name="parent_mesh")
    ```

3. **Initialize model object**: Once the SMART containers and the parent mesh have been defined, these can all be used to assemble a single `smart.model` object. In our example, we initialize the model object with the default configuration:
    ```python
    config_cur = smart.config.Config()
    model_cur = smart.model.Model(pc, sc, cc, rc, config_cur, parent_mesh)
    ```
4. **Specify solver configuration**: SMART solves the assembled reaction-transport equations using the finite element method. This finite element system is assembled using [FEniCS](https://fenicsproject.org/) and is then solved using [SNES in PETSc4Py](https://petsc.org/release/manual/snes/#ch-snes). The user can provide problem-tailored solver specifications by calling `config_cur.solver.update()` (given a config object, `config_cur`). Minimally, the user should provide a time step and final time for the simulation:
    ```python
    config_cur.solver.update({"final_t": 10.0, "initial_dt": 0.1})
    ```

>     In this manner, the user can also specify "time_precision" (number of digits past the decimal point to round time, default is 6) and "print_assembly" (prints information during assembly process, default is True).

Reviewer: I think we should remove the `print_assembly` option from Smart in its entirety.
As we are now using the logger for printing, I think it should only be the log-level that determines if we want to print or not, not a secondary parameter ("print_assembly"), that is passed to the `SmartSNESProblem` as the variable `verbose`. @finsberg  any thoughts on this?

Reviewer 2: I agree, I can make an issue for it
